exp/controllers/azureasomanagedcluster_controller.go

```diff
@@ -57,5 +77,81 @@ func (r *AzureASOManagedClusterReconciler) SetupWithManager(ctx context.Context,
 
 // Reconcile reconciles an AzureASOManagedCluster.
 func (r *AzureASOManagedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, resultErr error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx,
+		"controllers.AzureASOManagedClusterReconciler.Reconcile",
+		tele.KVP("namespace", req.Namespace),
+		tele.KVP("name", req.Name),
+		tele.KVP("kind", infrav1exp.AzureASOManagedClusterKind),
+	)
+	defer done()
+
+	asoManagedCluster := &infrav1exp.AzureASOManagedCluster{}
+	err := r.Get(ctx, req.NamespacedName, asoManagedCluster)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	patchHelper, err := patch.NewHelper(asoManagedCluster, r.Client)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create patch helper: %w", err)
+	}
+	defer func() {
+		err := patchHelper.Patch(ctx, asoManagedCluster)
+		if err != nil && resultErr == nil {
+			resultErr = err
+			result = ctrl.Result{}
+		}
+	}()
+
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, asoManagedCluster.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if cluster != nil && cluster.Spec.Paused ||
+		annotations.HasPaused(asoManagedCluster) {
+		return r.reconcilePaused(ctx, asoManagedCluster, cluster)
+	}
+
+	if !asoManagedCluster.GetDeletionTimestamp().IsZero() {
+		return r.reconcileDelete(ctx, asoManagedCluster)
+	}
+
+	return r.reconcileNormal(ctx, asoManagedCluster, cluster)
+}
+
+//nolint:unparam // these parameters will be used soon enough
+func (r *AzureASOManagedClusterReconciler) reconcileNormal(ctx context.Context, asoManagedCluster *infrav1exp.AzureASOManagedCluster, cluster *clusterv1.Cluster) (ctrl.Result, error) {
+	//nolint:all // ctx will be used soon
+	ctx, log, done := tele.StartSpanWithLogger(ctx,
+		"controllers.AzureASOManagedClusterReconciler.reconcileNormal",
+	)
+	defer done()
+	log.V(4).Info("reconciling normally")
+
+	return ctrl.Result{}, nil
+}
+
+//nolint:unparam // these parameters will be used soon enough
+func (r *AzureASOManagedClusterReconciler) reconcilePaused(ctx context.Context, asoManagedCluster *infrav1exp.AzureASOManagedCluster, cluster *clusterv1.Cluster) (ctrl.Result, error) {
+	//nolint:all // ctx will be used soon
+	ctx, log, done := tele.StartSpanWithLogger(ctx,
+		"controllers.AzureASOManagedClusterReconciler.reconcilePaused",
+	)
+	defer done()
+	log.V(4).Info("reconciling pause")
+
+	return ctrl.Result{}, nil
+}
+
+//nolint:unparam // these parameters will be used soon enough
+func (r *AzureASOManagedClusterReconciler) reconcileDelete(ctx context.Context, asoManagedCluster *infrav1exp.AzureASOManagedCluster) (ctrl.Result, error) {
+	//nolint:all // ctx will be used soon
```

Comment:
The `nolintlint` linter should be catching when these directives are no longer needed. And I've been trying to avoid adding TODOs because enumerating all the gaps at this point would make every PR like 50% TODO comments lol.

Reply:
There is a (convincing) school of thought that TODO comments should never be committed and there are linters to prove it, so IMO we would not do that as a norm on a prod codebase (it could be sane to grant an exception to something behind a feature flag, but I digress)

Reply:
These effectively are TODOs though, just without the literal text. I think tomorrow I'll refactor these to just not include the parameters in the first place so we don't need any lint directives.